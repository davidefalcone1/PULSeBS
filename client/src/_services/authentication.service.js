import { BehaviorSubject } from 'rxjs';


const currentUserSubject = new BehaviorSubject(JSON.parse(localStorage.getItem('user')));

export const authenticationService = {
    login,
    logout,
    user: currentUserSubject.asObservable(),
    get currentUserValue () { return currentUserSubject.value }
};

function handleResponse(response) {
    return response.text().then(text => {
        const data = text && JSON.parse(text);
        if (!response.ok) {
            if ([401, 403].indexOf(response.status) !== -1) {
                // auto logout if 401 Unauthorized or 403 Forbidden response returned from api
                authenticationService.logout();
            }

            const error = (data && data.message) || response.statusText;
            return Promise.reject(error);
        }

        return data;
    });
}

function login(username, password) {
    const requestOptions = {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ username, password })
    };
    return new Promise(async function (resolve, reject) {
        try {
            const response = await fetch('/users/authenticate', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify({ username, password }),
            });
            const user = await response.json().then(handleResponse);
            if (response.ok) {
                localStorage.setItem('user', JSON.stringify(user));
                currentUserSubject.next(user);
                resolve(user);
            }
            else
                reject();
        }
        catch (e) {
            reject();
        }
    });

}
function logout() {

    const requestOptions = {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
    };
    localStorage.removeItem('user');
    currentUserSubject.next(null);
    return fetch('/logout',requestOptions);
}