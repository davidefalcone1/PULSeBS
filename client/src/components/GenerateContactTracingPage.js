import React from 'react';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import Form from 'react-bootstrap/Form';
import Alert from 'react-bootstrap/Alert';
import {AuthContext} from '../_services/AuthContext';
import {Redirect} from 'react-router-dom';

const generateContactTracingPage = (props) => {
  return <GenerateContactTracing generateStudentTracing = {props.generateStudentTracing}
    generateTeacherTracing = {props.generateTeacherTracing}/>
}

class GenerateContactTracing extends React.Component{
    constructor(props){
        super(props);
        this.props = props;
        this.state = {
            studentID: '', teacherID: '',
            studentFormat: 'Select format', teacherFormat: 'Select format',
            errorStudent: false, errorTeacher: false,
            errorStudentFormat: false, errorTeacherFormat: false
        }
    }

    updateField = (name, value) => {
        this.setState({[name]: value}, () => {
            if(this.state.studentID !== ''){
              this.setState({errorStudent: false});
            }
            if(this.state.teacherID !== ''){
              this.setState({errorTeacher: false});
            }
            if(this.state.studentFormat !== 'Select format'){
                this.setState({errorStudentFormat: false});
            }
            if(this.state.teacherFormat !== 'Select format'){
                this.setState({errorTeacherFormat: false});
            }
        });
    }
    handleSubmitStudent = () => {
        if (!this.sForm.checkValidity()) {
            this.form.reportValidity();
        }
        else if(this.state.studentID === ''){
            this.setState({errorStudent: true});
        }
        else if(this.state.studentFormat === 'Select format'){
            this.setState({errorStudentFormat: true});
        }
        else {
            this.props.generateStudentTracing(this.state.studentID, this.state.studentFormat.toLowerCase())
        }
    }
    handleSubmitTeacher = () => {
        if (!this.sForm.checkValidity()) {
            this.form.reportValidity();
        }
        else if(this.state.teacherID === ''){
            this.setState({errorTeacher: true});
        }
        else if(this.state.teacherFormat !== 'Select format'){
            this.setState({errorTeacherFormat: true});
        }
        else {
            this.props.generateTeacherTracing(this.state.teacherID, this.state.teacherFormat.toLowerCase())
        }
    }

    render(){
        return(
            <div style={{padding: "15px"}}>
                <AuthContext.Consumer>
                {(context) => (
                    <>        
                    {context.user && 
                        <>
                            <Row>
                                <Col>
                                    <Form method="POST" action="" id="studentForm" onSubmit={(ev) => {
                                        ev.preventDefault();
                                        this.handleSubmitStudent();
                                    }} ref={(form) => this.sForm = form}>                          
                                        <Form.Group>
                                            <Form.Label className="control-label">Student Id</Form.Label>
                                            <Form.Control type="text" name="studentID" size = "lg"
                                            value = {this.state.studentID} required autoFocus
                                            onChange={(ev) => this.updateField(ev.target.name, ev.target.value)}/>
                                        </Form.Group>
                                        <Form.Group>
                                            <Form.Label className="control-label">Download Format</Form.Label>
                                            <Form.Control as="select" custom name="studentFormat" value = {this.state.studentFormat} 
                                                onChange={(ev) => this.updateField(ev.target.name, ev.target.value)}>
                                            <option>Select format</option>
                                            <option>PDF</option>
                                            <option>CSV</option>
                                            </Form.Control>
                                        </Form.Group>
                                        <Form.Group>
                                            <div>
                                                <button type="submit" className="btn btn-primary">Download</button>
                                                {this.state.errorStudent &&
                                                <>
                                                    <br/>
                                                    <Alert key="sError" variant="danger">
                                                        Student Id cannot be empty.
                                                    </Alert>
                                                </>}
                                                {this.state.errorStudentFormat &&
                                                <>
                                                    <br/>
                                                    <Alert key="tError" variant="danger">
                                                        You must select a valid format.
                                                    </Alert>
                                                </>}
                                            </div>
                                        </Form.Group>
                                    </Form>
                                </Col>
                                <Col>
                                    <Form method="POST" action="" id="teacherForm" onSubmit={(ev) => {
                                        ev.preventDefault();
                                        this.handleSubmitTeacher();
                                    }} ref={(form) => this.tForm = form}>                          
                                        <Form.Group>
                                            <Form.Label className="control-label">Teacher Id</Form.Label>
                                            <Form.Control type="text" name="teacherID" size = "lg"
                                            value = {this.state.teacherID} required autoFocus
                                            onChange={(ev) => this.updateField(ev.target.name, ev.target.value)}/>
                                        </Form.Group>
                                        <Form.Group>
                                            <Form.Label className="control-label">Download Format</Form.Label>
                                            <Form.Control as="select" custom name="teacherFormat" value = {this.state.teacherFormat} 
                                                onChange={(ev) => this.updateField(ev.target.name, ev.target.value)}>
                                            <option>Select format</option>
                                            <option>PDF</option>
                                            <option>CSV</option>
                                            </Form.Control>
                                        </Form.Group>
                                        <Form.Group>
                                            <div>
                                                <button type="submit" className="btn btn-primary">Download</button>
                                                {this.state.errorTeacher &&
                                                <>
                                                    <br/>
                                                    <Alert key="sError" variant="danger">
                                                        Teacher Id cannot be empty.
                                                    </Alert>
                                                </>}
                                                {this.state.errorTeacherFormat &&
                                                <>
                                                    <br/>
                                                    <Alert key="tError" variant="danger">
                                                        You must select a valid format.
                                                    </Alert>
                                                </>}
                                            </div>
                                        </Form.Group>
                                    </Form>
                                </Col>
                            </Row>
                        </>}
                    {!context.user && <Redirect to="/login"/>}
                </>
                )}
                </AuthContext.Consumer>
            </div>
        );
    }
}

export default generateContactTracingPage;
