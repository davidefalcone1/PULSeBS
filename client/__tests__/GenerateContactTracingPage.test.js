import React from 'react';
import renderer from 'react-test-renderer';
import { AuthContext } from '../src/_services/AuthContext';
import GenerateContactTracingPage from '../src/components/GenerateContactTracingPage';
import { BrowserRouter } from 'react-router-dom';

describe('GenerateContactTracingPage', ()=>{
    test('Renders correctly', ()=>{
        const component = renderer.create(
            <BrowserRouter>
                <AuthContext.Provider value={{}}>
                    <GenerateContactTracingPage generateStudentTracing={()=>{}} generateTeacherTracing={()=>{}}/>
                </AuthContext.Provider>
            </BrowserRouter>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});