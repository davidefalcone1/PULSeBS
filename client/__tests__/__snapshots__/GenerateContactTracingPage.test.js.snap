// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenerateContactTracingPage Renders correctly 1`] = `
<div
  style={
    Object {
      "padding": "15px",
    }
  }
/>
`;
